attempt to support windows

tamird opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/c-rocksdb/9)

<!-- Reviewable:end -->

tamird committed:
Preliminary Windows support

bdarnell commented:
How have you tested this? If we're going to attempt to support windows, appveyor.com is a good windows-based CI platform (and you can set up multiple CI systems with github; I do this for Tornado). 

However, this seems like an unnecessary distraction as we're approaching beta and we should wait to explore this until post-beta.

tamird commented:
Yes, I agree. I was trying to get cross-compilation working using https://github.com/karalabe/xgo, and wanted to document what I had tried. I won't be actively working on this.

tamird commented:
I should mention that this currently fails in xgo: https://github.com/karalabe/xgo/issues/26

petermattis commented on cgo_flags.go:
```diff
@@ -6,9 +6,10 @@ package rocksdb
+// #cgo CPPFLAGS: -DNDEBUG -DSNAPPY -DLZ4
+// #cgo darwin CPPFLAGS: -DOS_MACOSX -DROCKSDB_PLATFORM_POSIX
+// #cgo linux CPPFLAGS: -DOS_LINUX -DROCKSDB_PLATFORM_POSIX
+// #cgo windows CPPFLAGS: -DOS_WIN
```
Is someone asking for windows support? Just curious what motivated this.

tamird replied:
Nah, I was trying to get cross-compilation working so we could post binaries and wanted to see how far I could get with windows.

petermattis commented:
LGTM

tamird committed:
Windows: improve compatibility with mingw-w64

- Downcase some `#include`s since we run on case-sensitive file systems
- Avoid MSVC's `i64` suffix which isn't supported in GCC

tamird committed:
Windows: build without snappy to work around missing symbols

TL;DR: Portable Executables (Windows) don't support missing symbols
after link-time, and the go toolchain doesn't link dependencies into
cgo archives.

See https://github.com/golang/go/issues/14985.

tamird commented:
Hey, this is ready for another look. For now, I've worked around some issues by omitting snappy on windows. This compiles successfully with `xgo -go 1.6.1 --targets=windows-6.0/amd64 .`.

tbg commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but with a healthy dose of science dog.

<details><summary>

<i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>



> Hey, this is ready for another look. For now, I've worked around some issues by omitting snappy on windows. This compiles successfully with `xgo -go 1.6.1 --targets=windows-6.0/amd64 .`.
> </details>

---

Reviewed 6 of 8 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/c-rocksdb/9#-:-KIiIgDAEciGoq--UMy4:1173689132)_

<!-- Sent from Reviewable.io -->
